proposals/0000-ghc-module-naming.rst

```diff
@@ -0,0 +1,139 @@
+.. sectnum::
+
+**Module naming conventions for GHC base libraries**
+
+Background and motivation
+===========================
+The accepted `Proposal #51: GHC base libraries <https://github.com/haskellfoundation/tech-proposals/blob/main/proposals/accepted/051-ghc-base-libraries.rst>`_
+defines the following libraries:
+
+* ``base``: the foundational library on which the rest of the ecosystem is based.
+
+  * Its API is carefully curated by the `Core Libraries Committee <https://github.com/haskell/core-libraries-committee>`_, and is kept rather stable.
+
+* ``ghc-experimental``: the home of experimental extensions to GHC, usually ones proposed by the
+  `GHC Steering Committee <https://github.com/ghc-proposals/ghc-proposals/>`_.
+
+  * Functions and types in here are usually candidates for later transfer into ``base``.  But not necessarily: if a collection of functions is not adopted widely enough, it may not be proposed for a move to `base`.
+
+  * It is user-facing (user are encouraged to depend on it), but its API is less stable than ``base``.
+
+* ``ghc-prim, ghc-internals`` (and perhaps others): define functions and data types used internally by GHC to support the API of ``base`` and ``ghc-experimental``.
+
+  * These libraries come with no stability guarantees: they may change at short notice.
+
+In addition we already have:
+
+* ``ghc``: this library exposes GHC as a library, through the (currently ill-defined) GHC API.
+
+All these libraries follow the Haskell Package Versioning Policy (PVP).
+
+The question arises of *what module names should be used*. For example, suppose that all three exposed a module called ``Data.Tuple``.  In principle that would be fine -- GHC allows you
+to use the package name in the ``import`` statement, to disambiguate.  But it's *extremely* confusing.  This proposal articulates a set of conventions to
+help us design module names.
+
+The proposal
+============
+
+This proposal is split into four sub-proposals for easier discussion.  Each sub-proposal builds on the
+earlier ones -- they are increments, not alternatives.
+
+The goals of this proposal are deliberately limited to establish naming conventions.  We do not propose
+any changes to ``ghc`` or to ``cabal``.
+
+Proposal 1
+-----------
+
+* Modules in ``base``, ``ghc-experimental``, ``ghc-prim``, ``ghc-internals`` etc should all have distinct names.
+
+That principle leads immediately to the question: what should those names be?  Hence proposal 2.
+
+Proposal 2
+-----------
+
+* Modules in GHC's internal libraries (``ghc-prim``, ``ghc-internals`` etc) should be of form ``GHC.*``.
+* Modules in ``ghc-experimental`` should be of form ``Experimental.*``.
```

Comment:
I’d recommend to make the `Experiment` a _suffix_.

Rationale: `Data.Tuple.Experimental` is an companion/extension of `Data.Tuple`; some exports may move from one to the other. Many developers sort their imports alphabetically. Making this a suffix means all `Data.Tuple`-related imports are next to each other.

Ok (omitting explicit import lists and qualifiers):
```
import Control.Applicative
import Control.Arrow
import Experimental.Control.Applicative
import Experimental.Foreign.C
import Data.Tuple
import Foreign.C
```
Better:
```
import Control.Applicative
import Control.Applicative.Experimental
import Control.Arrow
import Data.Tuple
import Foreign.C
import Foreign.C.Experimental
```

Also, maybe people will use the idiom
```
import Data.List qualified as L
import Data.List.Experimental qualified as L
```
which is also nicer if both `qualified as L` are next to each other.

Reply:
Clearly this is a matter of taste.   Personally I find it easier to think of `Experimental.*` as a complete sub-tree of modules, all experimental.  And it's consistent with using a prefix `GHC.*` or `GhcAPI.*` elsewhere.  

But it doesn't matter what I feel provided whatever we do
* Satisfies the maximum number of users
* Is carried through consistently (e.g. all modules in `ghc-experimental` end in `.Experimental`.

I would love to hear from others about prefix-vs-postfix.

Reply:
I also think suffix is better. We've had very good luck not having package-distinct prefixes --- I am not sure why practice has gone better than theory! --- so I think it is OK to keep on doing that here.

Reply:
I think the reasoning for using a suffix is convincing.